lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -2661,75 +2659,207 @@ struct DotOpConversionHelper {
+  Value thread, lane, warp, warpMN, warpN, warpM;
+  size_t aElemBytes{}, bElemBytes{};
 
-  auto aShape = aTensorTy.getShape();
-  auto bShape = bTensorTy.getShape();
-  auto dShape = dTensorTy.getShape();
+  DotOpConversionHelper helper;
+  triton::DotOp op;
+  DotOpAdaptor adapter;
+  ConversionPatternRewriter &rewriter;
+  TypeConverter *typeConverter;
+  Location loc;
+  MLIRContext *ctx{};
 
-  auto mmaLayout = dTensorTy.getEncoding().cast<MmaEncodingAttr>();
+  using ValueTable = std::map<std::pair<unsigned, unsigned>, Value>;
+
+  MMA16816ConversionHelper(triton::DotOp op, Value thread, DotOpAdaptor adapter,
+                           ConversionPatternRewriter &rewriter,
+                           TypeConverter *typeConverter, Location loc)
+      : helper(op), op(op), adapter(adapter), rewriter(rewriter),
+        typeConverter(typeConverter), loc(loc), ctx(op.getContext()),
+        thread(thread) {
+    A = op.a();
+    B = op.b();
+    C = op.c();
+    D = op.c();
+
+    aTensorTy = A.getType().cast<RankedTensorType>();
+    bTensorTy = B.getType().cast<RankedTensorType>();
+    dTensorTy = D.getType().cast<RankedTensorType>();
+
+    aShape = aTensorTy.getShape();
+    bShape = bTensorTy.getShape();
+    dShape = dTensorTy.getShape();
+
+    mmaLayout = dTensorTy.getEncoding().cast<MmaEncodingAttr>();
+
+    wpt = mmaLayout.getWarpsPerCTA();
+
+    auto mmaInstrShape = helper.getMmaInstrShape();
+    mmaInstrM = mmaInstrShape[0];
+    mmaInstrN = mmaInstrShape[1];
+    mmaInstrK = mmaInstrShape[2];
+
+    auto matShape = helper.getMmaMatShape();
+    matShapeM = matShape[0];
+    matShapeN = matShape[1];
+    matShapeK = matShape[2];
+
+    int NK = aShape[1];
+    // shape / shape_per_cta
+    numRepM = std::max<int>(dShape[0] / (wpt[0] * mmaInstrM), 1);
+    numRepN = std::max<int>(dShape[1] / (wpt[1] * mmaInstrN), 1);
+    numRepK = std::max<int>(NK / mmaInstrK, 1);
+
+    Value _32 = i32_val(32);
+    lane = urem(thread, _32);
+    warp = udiv(thread, _32);
+    warpMN = udiv(warp, i32_val(wpt[0]));
+    warpM = urem(warp, i32_val(wpt[0]));
+    warpN = urem(warpMN, i32_val(wpt[1]));
+
+    aElemBytes = aTensorTy.getElementTypeBitWidth() / 8;
+    bElemBytes = bTensorTy.getElementTypeBitWidth() / 8;
+  }
+
+  // Loading $a from smem to registers, returns a LLVM::Struct.
+  Value loadA() {
+    ValueTable ha;
+    std::function<void(int, int)> loadFn;
+    if (aTensorTy.getEncoding().isa<SharedEncodingAttr>()) {
+      // load from smem
+      loadFn = getLoadMatrixFn(
+          A, adapter.a() /*llTensor*/, mmaLayout.getWarpsPerCTA()[0] /*wpt*/,
+          1 /*kOrder*/, {mmaInstrM, mmaInstrK} /*instrShpae*/,
+          {matShapeM, matShapeK} /*matShape*/, warpM /*warpId*/, ha /*vals*/);
+    } else if (aTensorTy.getEncoding().isa<BlockedEncodingAttr>()) {
+      // load from registers, used in gemm fuse
+      // TODO(Superjomn) Port the logic.
+      assert(false && "Loading A from register is not supported yet.");
```

Comment:
Nice todo:), since I think this logic is necessary for fusion pattern like fused multi-head attention.